### PR TITLE
Tonemapping: Added Luma-based Reinhard Tonemapping Option

### PR DESCRIPTION
**Pull request by DanielSturk**
**Description**

I propose the addition of Reinhard-Luminance tonemapping, from [this](https://64.github.io/tonemapping/) blog post. On that page, this is the function `reinhard_extended_luminance` with `max_white_l` set to infinity, and then the equation simplified. According to these sources ([1](https://en.wikipedia.org/wiki/Talk:Relative_luminance#Some_nitpicks_and_inaccuracies/confusions) [2](https://stackoverflow.com/questions/596216/formula-to-determine-perceived-brightness-of-rgb-color/56678483#56678483) ), the correct luminance coefficients for linear light are (0.2126, 0.7152, 0.0722), which is also the coefficients given in that blog post. The coefficients (0.299, 0.587, 0.114) which are pervasive in three.js are incorrect, but work better when the input is in sRGB/gamma space.

---------------------------------------------
This is a trivial improvement to the existing Reinhard tonemapping. The difference is that Reinhard-Luminance uses lumninance in the denominator, rather than curving the color channels independently. Curving the color channels independently can result in desaturation and incorrect hue (the hue veers towards cyan, yellow, or magenta). Reinhard-Luminance does the best job of not doing that.

| Exposure:  | 0.1 | 1 | 10 |
| ------------- | ------------- | ------------- | ------------- |
| Linear | image | image | image |
| Reinhard | image | image | image |
| Reinhard Luminance | image | image | image |
| Cineon | image | image | image |
| ACES | image | image | image |
---------------------------------------------
The existing tonemapping options quickly become saturated at sizeable exposure values. At low exposure values, Cineon and ACES are underexposed. Reinhard-Luminance doesn't have these issues.

| Exposure:  | 0.1 | 1 | 10 |
| ------------- | ------------- | ------------- | ------------- |
| Linear | image | image | image |
| Reinhard | image | image | image |
| Reinhard Luminance | image | image | image |
| Cineon | image | image | image |
| ACES | image | image | image |
---------------------------------------------
From my testing, some of the tonemapping options look good at certain brightnesses, but only Reinhard-Luminance looks consistently good at all brightnesses.
| Exposure:  | 0.01 | 0.1 | 1 | 4 |
| ------------- | ------------- | ------------- | ------------- | ------------- |
| Linear  | image | image | image | image |
| Reinhard  | image | image | image | image |
| Reinhard Luminance | image | image | image | image |
| Cineon  | image | image | image | image |
| ACES  | image | image | image | image |

----------------------------------------------

*This contribution is funded by [Threekit](https://www.threekit.com/)*

### CODE REVIEW

File: examples/webgl_loader_texture_exr.html [hunk 1]
```diff
@@ -33,8 +33,17 @@
...
+				Rienhard: THREE.ReinhardToneMapping,
+				RienhardLuma: THREE.ReinhardLumaToneMapping,
+				Cineon: THREE.CineonToneMapping,
+				ACES: THREE.ACESFilmicToneMapping,
+			};
+
 			const params = {
-				exposure: 2.0
+				logExposure: 0.0,
+				toneMapping: ToneMapping.Linear
```
Comment (WestLangley) [inline examples/webgl_loader_texture_exr.html:46]: I do not think we add tone mapping options to this example. Save that for the tone mapping examples.

File: examples/webgl_tonemapping.html [hunk 2]
```diff
@@ -67,7 +69,7 @@
 				document.body.appendChild( renderer.domElement );
 
 				renderer.toneMapping = toneMappingOptions[ params.toneMapping ];
-				renderer.toneMappingExposure = params.exposure;
+				renderer.toneMappingExposure = Math.pow( 10, params.logExposure );
```
Comment (WestLangley) [inline examples/webgl_tonemapping.html:72]: I do not think this re-parameterization of exposure is necessary. 

First, we do not want the user to set exposure to 10 -- much less 100. Such values are way over-bright.

Second, we may change exposure units to EVs, which is power-of-two based. But we have not had that discussion yet.

**Comment by WestLangley** [comment]
We can consider modifying the Reinhard tone mapping formula. We might not need both formulations, however.

**Comment by WestLangley** [comment]
>The coefficients (0.299, 0.587, 0.114) which are pervasive in three.js 

Sorry, but I do not see that as being a true statement.

File: examples/webgl_tonemapping.html (hunk 2 above)
Comment (DanielSturk) [reply examples/webgl_tonemapping.html:72]: Sure, I'll change it back. I needed it for the comparison images and figured others would want to be able to reproduce them

**Comment by WestLangley** [comment]
Be advised that three.js tone mapping does not "fix" an over-exposed scene. Exposure needs to first be set so middle-gray has the desired brightness, then tone mapping is applied.

It is not correct to make qualitative statements about which tone mapping operator is "best" when the scene is over-bright to begin with.

**Comment by DanielSturk** [comment]
Thank you for helping with this WestLangley

> We can consider modifying the Reinhard tone mapping formula. We might not need both formulations, however.

That would be much cleaner

> > The coefficients (0.299, 0.587, 0.114) which are pervasive in three.js
> 
> Sorry, but I do not see that as being a true statement.

To be clear, you're saying that (0.299, 0.587, 0.114) are more correct for linear luminance approximation than (0.2126, 0.7152, 0.0722)? (0.299, 0.587, 0.114) is used throughout three.js currently

> Be advised that three.js tone mapping does not "fix" an over-exposed scene. Exposure needs to first be set so middle-gray has the desired brightness, then tone mapping is applied.
Agreed... the high/low exposure settings I used were to produce a wider range of brightnesses for testing. By "looks good" I mean that the hue spectrum isn't replaced by CYM when mapping from a high dynamic range, and that some amount of saturation is preserved

**Comment by donmccurdy** [comment]
> Curving the color channels independently can result in desaturation and incorrect hue ... Reinhard-Luminance does the best job of not doing that.

Desaturation may also be [an intended effect of tone mapping](https://www.blenderguru.com/tutorials/secret-ingredient-photorealism#yui_3_17_2_1_1492511538692_15110), at the higher end of the dynamic range. I'd be cautious of claiming that less desaturation is necessarily best.

This addition seems to provide some flexibility that our existing options don't — particularly, less desaturation — so I'm good with the idea of including the option. I'm not sure I'd replace the existing Reinhard implementation, however.

Aside - at some point it could be nice to include `THREE.GrayscaleToneMapping` and `THREE.FalseColorToneMapping` options for debugging. These might help users to dial in the right exposure settings and identify over- or under-exposed areas.

**Comment by donmccurdy** [comment]
I should clarify — I agree there is room for improving and/or expanding the current tone mapping options. Thank you @DanielSturk for proposing this option, and for providing the test cases!

> I mean that the hue spectrum isn't replaced by CYM when mapping from a high dynamic range, and that some amount of saturation is preserved...

Reasonable goal, but honestly I'm not sure how to evaluate the quality of a particular tone mapping on this criteria. I think I'd expect a particular tone map to have a known dynamic range (ACES Filmic: `[-10 stops, +6.5 stops]`), to do as you suggest above for images properly exposed within that range, and to clamp outside that range. I don't know how to assess that today — perhaps we need more test cases, but I don't want to block the PR on them. If this is providing preferable results for properly-exposed scenes at Threekit, I would vote to include the option. Thanks again. 👍

**Comment by WestLangley** [comment]
>The coefficients (0.299, 0.587, 0.114) which are pervasive in three.js

>... (0.299, 0.587, 0.114) is used throughout three.js currently

Why do you say that?

Core provides
```glsl
float linearToRelativeLuminance( const in vec3 color ) {

	vec3 weights = vec3( 0.2126, 0.7152, 0.0722 );

	return dot( weights, color.rgb );

}
```

Methods in the examples that use sRGB-appropriate weights can be changed if necessary.

**Comment by WestLangley** [comment]
> If this is providing preferable results for properly-exposed scenes at Threekit, I would vote to include the option.

As I said, I would have no problem changing the Reinhard formula if a compelling case can be made that it is better.

**Comment by DanielSturk** [comment]
Reverted those demo changes. 

> > The coefficients (0.299, 0.587, 0.114) which are pervasive in three.js
> 
> > ... (0.299, 0.587, 0.114) is used throughout three.js currently
> 
> Why do you say that?
> 
> Core provides
> 
> ```glsl
> float linearToRelativeLuminance( const in vec3 color ) {
> 
> 	vec3 weights = vec3( 0.2126, 0.7152, 0.0722 );
> 
> 	return dot( weights, color.rgb );
> 
> }
> ```
> 
> Methods in the examples that use sRGB-appropriate weights can be changed if necessary.

Yes it's some examples that are using sRGB weights. Thanks for providing the core function, I missed that. Due to the way the includes are set up currently, I cannot call it from `tonemapping_pars_fragment.glsl` since that file is included before `common.glsl`. I copied the pattern that `saturate` was using, but could just revert to what I had if you prefer

**Comment by WestLangley** [comment]
@DanielSturk Thank you. I'll defer to others to decide if this implementation should replace the previous implementation of Reinhard.

**Comment by mrdoob** [comment]
>> If this is providing preferable results for properly-exposed scenes at Threekit, I would vote to include the option.
>
> As I said, I would have no problem changing the Reinhard formula if a compelling case can be made that it is better.

Agreed.

The current implementation of tonemapping makes it hard to to add options so we'd need a compelling case.

Are the any other users apart from Threekit interested in this?
If this is only for Threekit, we have the `THREE.CustomToneMapping` option for this.

**Comment by Mugen87** [comment]
To sum up, there are two options for a more specified version of Reinhard tone mapping.

1) Implement this on app level with `THREE.CustomToneMapping` for `WebGLRenderer` or with a custom node assigned to `WebGPURenderer.toneMappingNode`.
2) Improve the existing Reinhard tone mapping.

If @DanielSturk thinks option two is worth it, I would suggest to file a fresh PR that focus on the comparison of vanilla Reinhard with "extended" Reinhard.

**Pull request closed without merging**